File: src/MyMarina.Web/src/utils/priceResolver.ts
export interface PriceResolverPlan {
  transientRateKind: string | null | undefined;
  transientAmount: number | null | undefined;
  transientMinCharge: number | null | undefined;
  leaseRateKind: string | null | undefined;
  leaseAmount: number | null | undefined;
  leaseMinCharge: number | null | undefined;
  amenityAddOns: ReadonlyArray<{
    amenity: string;
    transientAmount: number | null | undefined;
    leaseAmount: number | null | undefined;
  }>;
}

export interface ResolvedPrices {
  transient: number | null;
  lease: number | null;
}

export function resolvePrice(
  plan: PriceResolverPlan,
  slipLength: number,
  slipBeam: number,
  amenities: readonly string[],
): ResolvedPrices {
  return {
    transient: resolveRate(
      plan.transientRateKind, plan.transientAmount, plan.transientMinCharge,
      slipLength, slipBeam, plan.amenityAddOns, amenities, 'transient',
    ),
    lease: resolveRate(
      plan.leaseRateKind, plan.leaseAmount, plan.leaseMinCharge,
      slipLength, slipBeam, plan.amenityAddOns, amenities, 'lease',
    ),
  };
}

function resolveRate(
  rateKind: string | null | undefined,
  amount: number | null | undefined,
  minCharge: number | null | undefined,
  length: number,
  beam: number,
  addOns: PriceResolverPlan['amenityAddOns'],
  amenities: readonly string[],
  side: 'transient' | 'lease',
): number | null {
  if (!rateKind || amount == null) return null;

  let base: number;
  if (rateKind === 'Flat') base = amount;
  else if (rateKind === 'PerFoot') base = amount * length;
  else if (rateKind === 'PerArea') base = amount * length * beam;
  else return null;

  for (const addOn of addOns) {
    if (!amenities.includes(addOn.amenity)) continue;
    const extra = side === 'transient' ? addOn.transientAmount : addOn.leaseAmount;
    if (extra != null) base += extra;
  }

  return Math.max(base, minCharge ?? 0);
}

File: src/MyMarina.Application/PricingPlans/PriceResolver.cs
using MyMarina.Domain.Entities;
using MyMarina.Domain.Enums;

namespace MyMarina.Application.PricingPlans;

public static class PriceResolver
{
    public static (decimal? Transient, decimal? Lease) Resolve(
        PricingPlan plan,
        decimal slipLength,
        decimal slipBeam,
        IEnumerable<PlanAmenity> slipAmenities)
    {
        var amenitySet = slipAmenities as ICollection<PlanAmenity> ?? slipAmenities.ToHashSet();

        return (
            Resolve(plan.TransientRateKind, plan.TransientAmount, plan.TransientMinCharge,
                    slipLength, slipBeam, amenitySet, transient: true, plan),
            Resolve(plan.LeaseRateKind, plan.LeaseAmount, plan.LeaseMinCharge,
                    slipLength, slipBeam, amenitySet, transient: false, plan)
        );
    }

    static decimal? Resolve(
        RateKind? rateKind,
        decimal? amount,
        decimal? minCharge,
        decimal slipLength,
        decimal slipBeam,
        ICollection<PlanAmenity> amenitySet,
        bool transient,
        PricingPlan plan)
    {
        if (rateKind is null || amount is null)
            return null;

        var baseRate = rateKind switch
        {
            RateKind.Flat    => amount.Value,
            RateKind.PerFoot => amount.Value * slipLength,
            RateKind.PerArea => amount.Value * slipLength * slipBeam,
            _                => throw new ArgumentOutOfRangeException(nameof(rateKind)),
        };

        var addOns = plan.AmenityAddOns
            .Where(a => amenitySet.Contains(a.Amenity))
            .Sum(a => transient ? (a.TransientAmount ?? 0m) : (a.LeaseAmount ?? 0m));

        var total = baseRate + addOns;
        return minCharge.HasValue ? Math.Max(total, minCharge.Value) : total;
    }

    public static IEnumerable<PlanAmenity> SlipAmenities(Domain.Entities.Slip slip)
    {
        if (slip.IsCovered) yield return PlanAmenity.Covered;
        if (slip.Electric == ElectricAmperage.Amp30) yield return PlanAmenity.Electric30A;
        if (slip.Electric == ElectricAmperage.Amp50) yield return PlanAmenity.Electric50A;
        if (slip.HasWater) yield return PlanAmenity.HasWater;
        if (slip.HasPumpOut) yield return PlanAmenity.HasPumpOut;
    }
}
